tests/bmtestspolar/symfixedrankpolarfactory.m

```diff
@@ -0,0 +1,264 @@
+	% convert a matrix to a pair, matrix2pair, allows to specify the
+	% rank of the representation. By default, it is equal to k. Omit the
+	% second input (or set to inf) to get a full EVD pair (in economy
+	% format). If so, the resulting pair does not represent a point on
+	% the manifold.
+	M.matrix2pair = @matrix2pair;
+	function X_pair = matrix2pair(X_matrix, r)
+	    if ~exist('r', 'var') || isempty(r) || r <= 0
+	        r = k;
+	    end
+	    if r < min(m, n)
+	        [U, B] = eigs(X_matrix, r);
+	    else
+	        [U, B] = eig(X_matrix);
+	    end
+	    X_pair.U = U;
+	    X_pair.B = B;
+	end
+	M.pair2matrix = @pair2matrix;
+	function X_matrix = pair2matrix(X_pair)
+	    U = X_pair.U;
+	    B = X_pair.B;
+	    X_matrix = U*B*U';
+	end
+    
+    M.egrad2rgrad = @egrad2rgrad;
+    function rgrad = egrad2rgrad(X, egrad)
+        rgrad.U = stiefel_proj(X.U, egrad.U);
+        rgrad.B = X.B*symm(egrad.B)*X.B;
+    end
+    
+    
+    M.ehess2rhess = @ehess2rhess;
+    function Hess = ehess2rhess(X, egrad, ehess, eta)
+        % Mathematical development of the below computation is
+        % as follows.
+        % We first compute the Riemannian gradient for the factor B.
+        % rgrad.B = X.B*symm(egrad.B)*X.B;
+        %
+        % Directional derivatives of the Riemannian gradient.
+        % Hess.U = ehess.U - eta.U*symm(X.U'*egrad.U);
+        % Hess.U = stiefel_proj(X.U, Hess.U);
+        %
+        % Hess.B = X.B*symm(ehess.B)*X.B +  2*symm(eta.B*symm(egrad.B)*X.B);
+        %
+        % Finally, a correction factor for the non-constant metric on the factor B.
+        % Hess.B = Hess.B - symm(eta.B*(X.B\rgrad.B));
+        Hess.U = ehess.U - eta.U*symm(X.U'*egrad.U);
+        Hess.U = stiefel_proj(X.U, Hess.U);
+
+        Hess.B = X.B*symm(ehess.B)*X.B + symm(eta.B*symm(egrad.B)*X.B);
+
+		
+        % Projection onto the horizontal space.
+        Hess = M.proj(X, Hess);
+    end
+    
+    
+    M.proj = @projection;
+    function etaproj = projection(X, eta)
+        % First, projection onto the tangent space of the total space.
+        eta.U = stiefel_proj(X.U, eta.U);
+        eta.B = symm(eta.B);
+        
+        % Then, projection onto the horizontal space.
+        SS = X.B*X.B';
+        AS = X.B*(skew(X.U'*eta.U) - 2*skew(X.B\eta.B))*X.B;
+        
+        % Compute skew-symmetric Omega.
+        % To solve the system SS*Omega + Omega*SS - B*Omega*B = AS
+        Omega = mylinearsystem(SS, AS);
+        
+        %         % Debug: Omega is skew-symmetric.
+        %         norm(Omega + Omega','fro')
+        
+        etaproj.U = eta.U - X.U*Omega;
+        etaproj.B = eta.B - (X.B*Omega - Omega*X.B);
+        
+        %         % Debug
+        %         neta.U = eta.U - etaproj.U;
+        %         neta.B = eta.B - etaproj.B;
+        %         M.inner(X, neta, etaproj)
+    end
+    
+    M.tangent = M.proj;
+    M.tangent2ambient = @(X, eta) eta;
```

Comment:
NB May 3, 2019 comment:

> I'm not so sure about tangent and tangent2ambient -- for such manifolds where points and vectors are represented through tuples, we have to be really clear about how point and vectors are represented in the ambient space and on the manifold. Is this really what we want here?

BM Sep 17, 2019 reply:

> This is okay, but I am not sure about the usefulness of a such a thing in this factory.

NB Sep 17, 2019 reply:

> What is the structure (or total, or "embedding") space? What is the representation format for points and vectors in the structure space? This should also be specified in the documentation of this factory at the top, because this tells the user how to supply egrad and ehess. This in turn tells us how to code proj, tangent and tangent2ambient.